Manage the list of DDC nodes

doasync opened the pull request:
### User Story
**As a** SC client
**I want** to be able to register and list all DDC nodes
**So that** I can do it using SC functions

### New functions
- `add_ddc_node(id: String, url: String)`
- `get_all_ddc_nodes()`

### Tasks
[CBI-781](https://cerenetwork.atlassian.net/browse/CBI-781):
- [CBI-782](https://cerenetwork.atlassian.net/browse/CBI-782)
- [CBI-783](https://cerenetwork.atlassian.net/browse/CBI-783)

evgeny-s commented on cere02/lib.rs:
```diff
@@ -36,6 +36,7 @@ mod ddc {
         // -- Metrics Reporting --
         pub metrics: StorageHashMap<MetricKey, MetricValue>,
         current_period_ms: u64,
+        ddn_nodes: Vec<DDNNode>
```
Either change Name to DDCNode or just DDN.

doasync replied:
I assumed DDN stands for "decentralized data network" ([source](https://blog.cere.network/cere-is-here-the-future-of-customer-data/)). Yeah, let's call them DDC nodes.

doasync commented:
Do we need an event `DDCNodeAdded`?

naure commented on cere02/lib.rs:
```diff
@@ -638,6 +640,35 @@ mod ddc {
+    #[derive(Default, Clone, PartialEq, Eq, PartialOrd, Ord, Encode, Decode, SpreadLayout, PackedLayout)]
+    #[cfg_attr(feature = "std", derive(Debug, scale_info::TypeInfo))]
+    pub struct DDCNode {
+        id: String,
```
Please rename everywhere `id` to `p2p_id`, to not confuse with account ID. (refers to [libp2p](https://docs.libp2p.io/concepts/peer-id/))

doasync replied:
done

naure commented on cere02/lib.rs:
```diff
@@ -638,6 +640,35 @@ mod ddc {
+        #[ink(message)]
+        pub fn add_ddc_node(&mut self, id: String, url: String) -> Result<()> {
+            let reporter = self.env().caller();
+            self.only_reporter(&reporter)?;
```
This should be managed by the admin for now, not reporters.

doasync replied:
done

evgeny-s commented:
> Do we need an event DDCNodeAdded?
Yes, let's keep it

naure reviewed (changes requested):
Nice work.

Let’s complete it right away with:
- Change to only_admin
- `remove_` and `is_` functions.
- Events for Add / Remove.

Basically the same as for reporters.
You may also move the code just after the "Admin: Reporters" section to show that it is similar.

doasync commented:
@naure Thanks for your comments. I have updated the code:
- implemented emitting of add/remove events
- implemented `remove_` and `is_` functions
- set only_owner for add/remove
- renamed ddc node id to p2p_id
- moved ddc nodes code after the reporters
- refactored code

naure commented on cere02/lib.rs:
```diff
@@ -488,6 +492,64 @@ mod ddc {
+    #[ink(event)]
+    pub struct DDCNodeAdded {
+        #[ink(topic)]
+        p2p_id: String,
```
There should be `url` too.

doasync replied:
[done](https://github.com/Cerebellum-Network/cere-enterprise-smart-contracts/pull/20/files#diff-a99b45b57bfdc35f8b28ef154fa4418c459a459aa60dfe0f98d8a350a93fa067R508)

naure commented on cere02/lib.rs:
```diff
@@ -949,6 +1010,11 @@ mod ddc {
             assert_eq!(contract.get_current_period_ms(), today_ms);
         }
 
+        fn decode_event(event: &ink_env::test::EmittedEvent) -> Event {
```
Good idea.

naure reviewed (approved):
Just one more request.

Great job.
